// payload/dist/admin/components/views/Global/Default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DefaultGlobalEdit", {
    enumerable: true,
    get: function() {
        return DefaultGlobalEdit;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../../utilities/getTranslation");
const _DocumentControls = require("../../../elements/DocumentControls");
const _DocumentFields = require("../../../elements/DocumentFields");
const _LeaveWithoutSaving = require("../../../modals/LeaveWithoutSaving");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../../utilities/Meta"));
const _SetStepNav = require("../../collections/Edit/SetStepNav");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const DefaultGlobalEdit = (props)=>{
    const { apiURL, data, fieldTypes, global, permissions } = props;
    const { i18n } = (0, _reacti18next.useTranslation)();
    const { admin: { description, forceRenderAllFields } = {}, fields, label } = global;
    const hasSavePermission = permissions?.update?.permission;
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: (0, _getTranslation.getTranslation)(label, i18n),
        keywords: `${(0, _getTranslation.getTranslation)(label, i18n)}, Payload, CMS`,
        title: (0, _getTranslation.getTranslation)(label, i18n)
    }), !(global.versions?.drafts && global.versions?.drafts?.autosave) && /*#__PURE__*/ _react.default.createElement(_LeaveWithoutSaving.LeaveWithoutSaving, null), /*#__PURE__*/ _react.default.createElement(_SetStepNav.SetStepNav, {
        global: global
    }), /*#__PURE__*/ _react.default.createElement(_DocumentControls.DocumentControls, {
        apiURL: apiURL,
        data: data,
        global: global,
        hasSavePermission: hasSavePermission,
        isEditing: true,
        permissions: permissions
    }), /*#__PURE__*/ _react.default.createElement(_DocumentFields.DocumentFields, {
        description: description,
        fieldTypes: fieldTypes,
        fields: fields,
        forceRenderAllFields: forceRenderAllFields,
        hasSavePermission: hasSavePermission,
        permissions: permissions
    }));
};
