// payload/dist/admin/components/elements/DocumentFields/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DocumentFields", {
    enumerable: true,
    get: function() {
        return DocumentFields;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _RenderFields = /*#__PURE__*/ _interop_require_default(require("../../forms/RenderFields"));
const _filterFields = require("../../forms/RenderFields/filterFields");
const _OperationProvider = require("../../utilities/OperationProvider");
const _Gutter = require("../Gutter");
const _ViewDescription = /*#__PURE__*/ _interop_require_default(require("../ViewDescription"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'document-fields';
const DocumentFields = (props)=>{
    const { AfterFields, BeforeFields, description, fieldTypes, fields, forceRenderAllFields, forceSidebarWrap, hasSavePermission, permissions } = props;
    const operation = (0, _OperationProvider.useOperation)();
    const sidebarFields = (0, _filterFields.filterFields)({
        fieldSchema: fields,
        fieldTypes,
        filter: (field)=>field?.admin?.position === 'sidebar',
        operation,
        permissions: permissions.fields,
        readOnly: !hasSavePermission
    });
    const hasSidebarFields = sidebarFields && sidebarFields.length > 0;
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            hasSidebarFields ? `${baseClass}--has-sidebar` : `${baseClass}--no-sidebar`,
            forceSidebarWrap && `${baseClass}--force-sidebar-wrap`
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__main`
    }, /*#__PURE__*/ _react.default.createElement(_Gutter.Gutter, {
        className: `${baseClass}__edit`
    }, /*#__PURE__*/ _react.default.createElement("header", {
        className: `${baseClass}__header`
    }, description && /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__sub-header`
    }, /*#__PURE__*/ _react.default.createElement(_ViewDescription.default, {
        description: description
    }))), BeforeFields || null, /*#__PURE__*/ _react.default.createElement(_RenderFields.default, {
        className: `${baseClass}__fields`,
        fieldSchema: fields,
        fieldTypes: fieldTypes,
        filter: (field)=>!field.admin.position || field.admin.position && field.admin.position !== 'sidebar',
        forceRenderAllFields: forceRenderAllFields,
        permissions: permissions.fields,
        readOnly: !hasSavePermission
    }), AfterFields || null)), hasSidebarFields && /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__sidebar-wrap`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__sidebar`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__sidebar-fields`
    }, /*#__PURE__*/ _react.default.createElement(_RenderFields.default, {
        fieldTypes: fieldTypes,
        fields: sidebarFields,
        forceRenderAllFields: forceRenderAllFields,
        permissions: permissions.fields,
        readOnly: !hasSavePermission
    }))))));
};
