// payload/dist/versions/getLatestCollectionVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getLatestCollectionVersion", {
    enumerable: true,
    get: function() {
        return getLatestCollectionVersion;
    }
});
const _combineQueries = require("../database/combineQueries");
const _types = require("../types");
const _appendVersionToQueryKey = require("./drafts/appendVersionToQueryKey");
const getLatestCollectionVersion = async ({ id, config, payload, query, req })=>{
    let latestVersion;
    const hasConfigDb = Object.keys(config?.db ? config?.db : {}).length > 0;
    if (config.versions?.drafts && !hasConfigDb) {
        const { docs } = await payload.db.findVersions({
            collection: config.slug,
            limit: 1,
            pagination: false,
            req,
            sort: '-updatedAt',
            where: (0, _combineQueries.combineQueries)((0, _appendVersionToQueryKey.appendVersionToQueryKey)(query.where), {
                parent: {
                    equals: id
                }
            })
        });
        [latestVersion] = docs;
    }
    let doc;
    if (config?.db?.findOne) {
        doc = await config.db.findOne({
            ...query,
            req
        });
    } else {
        doc = await payload.db.findOne({
            ...query,
            req
        });
    }
    if (!latestVersion || (0, _types.docHasTimestamps)(doc) && latestVersion.updatedAt < doc.updatedAt) {
        return doc;
    }
    return {
        ...latestVersion.version,
        id,
        createdAt: latestVersion.createdAt,
        updatedAt: latestVersion.updatedAt
    };
};
