// payload/dist/admin/components/views/Versions/columns.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "buildVersionColumns", {
    enumerable: true,
    get: function() {
        return buildVersionColumns;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _formatDate = require("../../../utilities/formatDate");
const _SortColumn = /*#__PURE__*/ _interop_require_default(require("../../elements/SortColumn"));
const _Config = require("../../utilities/Config");
const _AutosaveCell = require("./cells/AutosaveCell");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const CreatedAtCell = ({ id, collection, date, global })=>{
    const { admin: { dateFormat }, routes: { admin } } = (0, _Config.useConfig)();
    const { params: { id: docID } } = (0, _reactrouterdom.useRouteMatch)();
    const { i18n } = (0, _reacti18next.useTranslation)();
    let to;
    if (collection) to = `${admin}/collections/${collection.slug}/${docID}/versions/${id}`;
    if (global) to = `${admin}/globals/${global.slug}/versions/${id}`;
    return /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
        to: to
    }, date && (0, _formatDate.formatDate)(date, dateFormat, i18n?.language));
};
const TextCell = ({ children })=>/*#__PURE__*/ _react.default.createElement("span", null, children);
const buildVersionColumns = (collection, global, t, latestDraftVersion, latestPublishedVersion)=>{
    const entityConfig = collection || global;
    const columns = [
        {
            name: '',
            accessor: 'updatedAt',
            active: true,
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SortColumn.default, {
                    label: t('general:updatedAt'),
                    name: "updatedAt"
                }),
                renderCell: (row, data)=>/*#__PURE__*/ _react.default.createElement(CreatedAtCell, {
                        collection: collection,
                        date: data,
                        global: global,
                        id: row?.id
                    })
            },
            label: ''
        },
        {
            name: '',
            accessor: 'id',
            active: true,
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SortColumn.default, {
                    disable: true,
                    label: t('versionID'),
                    name: "id"
                }),
                renderCell: (row, data)=>/*#__PURE__*/ _react.default.createElement(TextCell, null, data)
            },
            label: ''
        }
    ];
    if (entityConfig?.versions?.drafts || entityConfig?.versions?.drafts && entityConfig.versions.drafts?.autosave) {
        columns.push({
            name: '',
            accessor: 'autosave',
            active: true,
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SortColumn.default, {
                    disable: true,
                    label: t('status'),
                    name: "autosave"
                }),
                renderCell: (row)=>{
                    return /*#__PURE__*/ _react.default.createElement(_AutosaveCell.AutosaveCell, {
                        latestDraftVersion: latestDraftVersion,
                        latestPublishedVersion: latestPublishedVersion,
                        rowData: row
                    });
                }
            },
            label: ''
        });
    }
    return columns;
};
