// payload/dist/config/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "defaults", {
    enumerable: true,
    get: function() {
        return defaults;
    }
});
const _path = /*#__PURE__*/ _interop_require_default(require("path"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const defaults = {
    admin: {
        avatar: 'default',
        buildPath: _path.default.resolve(process.cwd(), './build'),
        components: {},
        css: _path.default.resolve(__dirname, '../admin/scss/custom.css'),
        dateFormat: 'MMMM do yyyy, h:mm a',
        disable: false,
        inactivityRoute: '/logout-inactivity',
        indexHTML: _path.default.resolve(__dirname, '../admin/index.html'),
        logoutRoute: '/logout',
        meta: {
            titleSuffix: '- Payload'
        }
    },
    collections: [],
    cookiePrefix: 'payload',
    cors: [],
    csrf: [],
    custom: {},
    defaultDepth: 2,
    defaultMaxTextLength: 40000,
    endpoints: [],
    express: {
        compression: {},
        json: {},
        middleware: [],
        postMiddleware: [],
        preMiddleware: []
    },
    globals: [],
    graphQL: {
        disablePlaygroundInProduction: true,
        maxComplexity: 1000,
        schemaOutputFile: `${typeof process?.cwd === 'function' ? process.cwd() : ''}/schema.graphql`
    },
    hooks: {},
    joiValidation: true,
    localization: false,
    maxDepth: 10,
    rateLimit: {
        max: 500,
        window: 15 * 60 * 1000
    },
    routes: {
        admin: '/admin',
        api: '/api',
        graphQL: '/graphql',
        graphQLPlayground: '/graphql-playground'
    },
    serverURL: '',
    telemetry: true,
    typescript: {
        outputFile: `${typeof process?.cwd === 'function' ? process.cwd() : ''}/payload-types.ts`
    },
    upload: {}
};
