// payload/dist/admin/components/views/Global/Default.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../utilities/getTranslation");
const _DocumentHeader = require("../../elements/DocumentHeader");
const _Loading = require("../../elements/Loading");
const _Form = /*#__PURE__*/ _interop_require_default(require("../../forms/Form"));
const _ActionsProvider = require("../../utilities/ActionsProvider");
const _OperationProvider = require("../../utilities/OperationProvider");
const _SetStepNav = require("../collections/Edit/SetStepNav");
const _Routes = require("./Routes");
const _CustomComponent = require("./Routes/CustomComponent");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'global-edit';
const DefaultGlobalView = (props)=>{
    const { i18n } = (0, _reacti18next.useTranslation)('general');
    const { action, apiURL, data, disableRoutes, fieldTypes, global, initialState, isLoading, onSave, permissions } = props;
    const { setViewActions } = (0, _ActionsProvider.useActions)();
    const { label } = global;
    const hasSavePermission = permissions?.update?.permission;
    (0, _react.useEffect)(()=>{
        const path = location.pathname;
        if (!path.endsWith(global.slug)) {
            return;
        }
        const editConfig = global?.admin?.components?.views?.Edit;
        const defaultActions = editConfig && 'Default' in editConfig && 'actions' in editConfig.Default ? editConfig.Default.actions : [];
        setViewActions(defaultActions);
        return ()=>{
            setViewActions([]);
        };
    }, [
        global.slug,
        location.pathname,
        global?.admin?.components?.views?.Edit,
        setViewActions
    ]);
    return /*#__PURE__*/ _react.default.createElement("main", {
        className: `${baseClass} ${baseClass}--${global.slug}`
    }, /*#__PURE__*/ _react.default.createElement(_OperationProvider.OperationContext.Provider, {
        value: "update"
    }, /*#__PURE__*/ _react.default.createElement(_SetStepNav.SetStepNav, {
        global: global
    }), /*#__PURE__*/ _react.default.createElement(_Form.default, {
        action: action,
        className: `${baseClass}__form`,
        disabled: !hasSavePermission,
        initialState: initialState,
        method: "post",
        onSuccess: onSave
    }, /*#__PURE__*/ _react.default.createElement(_Loading.FormLoadingOverlayToggle, {
        action: "update",
        loadingSuffix: (0, _getTranslation.getTranslation)(label, i18n),
        name: `global-edit--${typeof label === 'string' ? label : label?.en}`
    }), !isLoading && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_DocumentHeader.DocumentHeader, {
        apiURL: apiURL,
        data: data,
        global: global
    }), disableRoutes ? /*#__PURE__*/ _react.default.createElement(_CustomComponent.CustomGlobalComponent, {
        view: "Default",
        ...props
    }) : /*#__PURE__*/ _react.default.createElement(_Routes.GlobalRoutes, {
        ...props,
        fieldTypes: fieldTypes
    })))));
};
const _default = DefaultGlobalView;
